Add Flashcards Backend

rliu13 opened the pull request:
## This pull request addresses issue #37 by creating the backend for the flashcards application.

- Implemented database connection and CRUD operations within dbfuncs
- Added flashcard and category pydantic models 
- Updated paths and app.py to include routing for my requests

srstone20 commented on backend/src/dbfuncs/flashcards.py:
```diff
@@ -0,0 +1,57 @@
+
+# Delete Flashcard
+async def delete_flashcard(flashcard_id: str, user_email: str):
+    result = await collection.delete_one({"question": flashcard_id, "user_email": user_email})
```
Should this function delete by id or question? It's accepting the "flashcard_id" but deleting by "question". I think you should **delete by ID**; multiple flashcards could have the same question in different categories, meaning more than one can get deleted on accident

srstone20 commented on backend/src/dbfuncs/flashcards.py:
```diff
@@ -0,0 +1,57 @@
+# Update Flashcard
+async def update_flashcard(flashcard_question: str, updated_flashcard: Flashcard, user_email: str):
+    result = await collection.update_one(
+        {"question": flashcard_question, "user_email": user_email},
```
PUT/updating should use flashcard_id, too. Similar reason to above

srstone20 commented on backend/src/paths/app/flashcards.py:
```diff
@@ -0,0 +1,42 @@
+    
+# Update Flashcard
+@router.put("/flashcards/{flashcard_question}", tags=["flashcards"])
+async def put_flashcard(flashcard_question: str, updated_flashcard: Flashcard, user_email: str):
```
In the current state, you can update the email. Please remove this because we don't want "ownership" of flashcards to change like this

srstone20 replied:
You probably need to make a different model to achieve this

srstone20 commented on backend/src/paths/app/categories.py:
```diff
@@ -0,0 +1,42 @@
+    
+# Update Category
+@router.put("/categories/{category_name}", tags=["flashcards"])
+async def put_category(category_name: str, updated_category: Category, user_email: str):
```
In the current state, you can update the email. Please remove this because we don't want "ownership" of categories to change like this. There is a similar issue with flashcards (see below)

srstone20 commented on backend/src/paths/app/categories.py:
```diff
@@ -0,0 +1,42 @@
+
+# Delete Category
+@router.delete("/categories/{category_name}", tags=["flashcards"])
+async def delete_category_route(category_name: str, user_email: str):
```
It's okay to delete by category name because they are unique, so keep this

rliu13 commented:
I updated the flashcard model to include an id property and updated all the problematic flashcard functions so that they use the _id property rather than user email or the question itself. I decided to remove the category update functionality for now until we can get user authentication implemented.

srstone20 commented:
Approved! Very nice work!
